### PR TITLE
fix data race in ppm #309

### PR DESCRIPTION
**Pull request by Grimnir777**
Signed-off-by: Thibaud Germain <thibaud.germain1@ibm.com>

### CODE REVIEW

File: token/core/zkatdlog/crypto/ppm/ppm.go [hunk 1]
```diff
@@ -59,6 +65,9 @@ func (v *PublicParamsManager) ForceFetch() error {
 }
 
 func (v *PublicParamsManager) PublicParams() *crypto.PublicParams {
+	v.mutex.Lock()
```
Comment (adecaro) [inline token/core/zkatdlog/crypto/ppm/ppm.go:68]: What about using first a read lock to check if v.pp is not nil and then a full lock in the case the public params must be fetched?

**Review by adecaro — CHANGES_REQUESTED** [review]

**Comment by adecaro** [comment]
@Grimnir777 , thanks for submitting this PR. 
Please, could you open a github issue that describes the problem?
Also, does this apply also to the fabtoken driver? Shall we apply the same patch there as well?

**Comment by Grimnir777** [comment]
Well we didn't see any issues in the fabtoken driver but for sure I can apply the changes also there

File: token/core/zkatdlog/crypto/ppm/ppm.go (hunk 1 above)
Comment (Grimnir777) [reply token/core/zkatdlog/crypto/ppm/ppm.go:68]: Is this working ? I'm not really familiar with the RWmutex 

```
	v.mutex.RLock()
	defer v.mutex.RUnlock()
	if v.pp == nil {
		v.mutex.Lock()
		defer v.mutex.Unlock()
		if v.publicParamsLoader == nil {
			panic("public parameters loaded not set")
		}
		var err error
		v.pp, err = v.publicParamsLoader.Load()
		if err != nil {
			panic(err)
		}
	}
```

Comment (adecaro) [reply token/core/zkatdlog/crypto/ppm/ppm.go:68]: What about this?

```go	
        v.mutex.RLock()
	if v.pp != nil {
		v.mutex.RUnlock()
		return v.pp
	}
	v.mutex.RUnlock()

	v.mutex.Lock()
	defer v.mutex.Unlock()
	if v.pp == nil {
		if v.publicParamsLoader == nil {
			panic("public parameters loaded not set")
		}
		var err error
		v.pp, err = v.publicParamsLoader.Load()
		if err != nil {
			panic(err)
		}
	}
	return v.pp
```

**Comment by adecaro** [comment]
Yes, please, can you apply the change also to fabtoken? Thanks :)

**Comment by adecaro** [comment]
Hi @Grimnir777, I think we need a simpler locking strategy now.
A PR merged yesterday simplified the public parameter manager.
Can you rebase, please?

**Comment by Grimnir777** [comment]
Hello @adecaro I did the update from the rebase

**Comment by adecaro** [comment]
Hi @Grimnir777 , thanks for updating the PR.
One of checks fails:

```
The following files contain goimports errors
token/core/fabtoken/ppm.go
make: *** [checks.mk:25: goimports] Error 1
The goimports command 'goimports -l -w' must be run for these files```
```

If you want to test locally on your machine, you can run `make checks`. 
Thanks :)

**Comment by adecaro** [comment]
@Grimnir777 , thanks a lot. I'll merge as soon as the CI passes :)

**Review by adecaro — APPROVED** [review]

**Pull request merged**